### ISSUE TITLE
Transition to TCP CLOSED is dropping packet

### ISSUE THREAD

**Issue by rzezeski** [oxidecomputer/opte#174]
When transition a TCP flow to CLOSED we are currently dropping the final packet.

https://github.com/oxidecomputer/opte/issues/129#issuecomment-1106645342

**Comment by rzezeski** [comment]
One way to test this is to verify that RST segments make it out. Previous to this fix all RST segments are dropped. This can be proven by trying to SSH into a guest on the wrong port.

* SSH from laptop to guest on port 7777.

```
rpz@sector5 [~]
$ ssh -o StrictHostKeyChecking=no -o ControlMaster=no -p 7777 root@10.0.0.226
^C
```

* Traffic from guest perspective is that it received SYN for bogus port, and replies with RST.

```
rpz@kalm:~$ pfexec snoop -r -d vopte0 tcp
Using device vopte0 (promiscuous mode)
   10.0.0.10 -> 172.30.0.5   TCP D=7777 S=57612 Syn Seq=3454304208 Len=0 Win=65535 Options=<mss 1460,nop,wscale 6,nop,nop,tstamp 621109780 0,sackOK,eol>
  172.30.0.5 -> 10.0.0.10    TCP D=57612 S=7777 Rst Ack=3454304209 Win=0
```

* Traffic from host ("below" OPTE) is that SYN segments continue to be retransmitted, but RST is never seen from guest.

```
rpz@kalm:~$ pfexec snoop -r -d net0 'tcp and ((src 10.0.0.226) or (dst 10.0.0.226))'
Using device net0 (promiscuous mode)
   10.0.0.10 -> 10.0.0.226   TCP D=7777 S=57613 Syn Seq=2415313140 Len=0 Win=65535 Options=<mss 1460,nop,wscale 6,nop,nop,tstamp 3347029712 0,sackOK,eol>
   10.0.0.10 -> 10.0.0.226   TCP D=7777 S=57613 Syn Seq=2415313140 Len=0 Win=65535 Options=<mss 1460,nop,wscale 6,nop,nop,tstamp 3347030713 0,sackOK,eol>
   10.0.0.10 -> 10.0.0.226   TCP D=7777 S=57613 Syn Seq=2415313140 Len=0 Win=65535 Options=<mss 1460,nop,wscale 6,nop,nop,tstamp 3347031715 0,sackOK,eol>
```

* This is because RST is dropped and causes connection to enter closed state at which point all traffic is dropped.

```
rpz@kalm:~/opte-dtrace$ pfexec dtrace -n 'tcp-flow-drop'
dtrace: description 'tcp-flow-drop' matched 2 probes
CPU     ID                    FUNCTION:NAME
  1  12213 _ZN4opte6engine4port4Port7process17hecebe1beb9b8488aE:tcp-flow-drop 
  1  12213 _ZN4opte6engine4port4Port7process17hecebe1beb9b8488aE:tcp-flow-drop 
  1  12213 _ZN4opte6engine4port4Port7process17hecebe1beb9b8488aE:tcp-flow-drop 
  1  12213 _ZN4opte6engine4port4Port7process17hecebe1beb9b8488aE:tcp-flow-drop 

rpz@kalm:~/oxidecomputer/opte/opteadm$ pfexec opteadm dump-tcp-flows -p opte0
TCP:172.30.0.5:7777:10.0.0.10:57612 FlowEntryDump { hits: 0, state_summary: "TCP:10.0.0.10:57612:10.0.0.226:7777 CLOSED Some(0) Some(3454304209) Some(3454304208) None" }
TCP:172.30.0.5:7777:10.0.0.10:57613 FlowEntryDump { hits: 0, state_summary: "TCP:10.0.0.10:57613:10.0.0.226:7777 CLOSED Some(0) Some(2415313141) Some(2415313140) None" }
TCP:172.30.0.5:7777:10.0.0.10:57614 FlowEntryDump { hits: 0, state_summary: "TCP:10.0.0.10:57614:10.0.0.226:7777 CLOSED Some(0) Some(446697128) Some(446697127) None" }
```

**Comment by rzezeski** [comment]
This work stalled a bit because the RST from the guest was not actually making it out even though the TCP flow was closed/removed, resulting in panics induced by a retrans of the SYN causing a UFT hit but no corresponding TCP flow table entry. This is a result of the fact that TCP flow tracking does not currently feed into UFT invalidation (#3). Hooking this up resulted in a bit of churn, but now an RST can finally make it back out to the client.

* Client seeing conn refused for SSH to bogus port.

```
rpz@sector5 [~]
$ ssh -o StrictHostKeyChecking=no -o ControlMaster=no -p 7777 root@10.0.0.226
ssh: connect to host 10.0.0.226 port 7777: Connection refused
```

* Host NIC seeing RST coming outbound from OPTE to be delivered to the client.

```
rpz@kalm:~$ pfexec snoop -r -d net0 'tcp and (src 10.0.0.226) or (dst 10.0.0.226)'
Using device net0 (promiscuous mode)
   10.0.0.10 -> 10.0.0.226   TCP D=7777 S=57964 Syn Seq=3147141055 Len=0 Win=65535 Options=<mss 1460,nop,wscale 6,nop,nop,tstamp 837185385 0,sackOK,eol>
  10.0.0.226 -> 10.0.0.10    TCP D=57964 S=7777 Rst Ack=3147141056 Win=0
```

**Comment by rzezeski** [comment]
Along with this work will come generating RSTs when unexpected TCP segments come in, just like an OS stack would do. That work will address the issue seen in #179.

**Comment by rzezeski** [comment]
> Along with this work will come generating RSTs when unexpected TCP segments come in, just like an OS stack would do. That work will address the issue seen in #179.

As I said in #179, I decided that OPTE shouldn't be generating resets on the guest's behalf.

**Comment by rzezeski** [comment]
Addressed in fb01edd13197fbc8626f852397ce2090a8502cbe.
